File: examples/c++/src/services/storage.cpp
#include <string>
#include <map>
#include "../temp_libs/json.hpp"
#include "../client.hpp"
#include "../exception.hpp"
#include "../service.hpp"
#include "storage.hpp"

using string = std::string;
using json = nlohmann::json;

namespace Appwrite {
/*
 * List Files
 *
     * Get a list of all the user files. You can use the query params to filter
     * your results. On admin mode, this endpoint will return a list of all of the
     * project's files. [Learn more about different API modes](/docs/admin).
 * @param string search
 * @param int limit
 * @param int offset
 * @param string orderType
 * @throws AppwriteException
 * @return array
 */
json Storage::listFiles(string search, int limit, int offset, string orderType) {
    json params = {
        {"search", search},
        {"limit", limit},
        {"offset", offset},
        {"orderType", orderType},
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files";
    size_t pos;
    string searchString;

    return this->client->call(Client::METHOD_GET, path, headers, params);
}

/*
 * Create File
 *
     * Create a new file. The user who creates the file will automatically be
     * assigned to read and write access unless he has passed custom values for
     * read and write arguments.
 * @param file file
 * @param array read
 * @param array write
 * @throws AppwriteException
 * @return array
 */
json Storage::createFile(file file, array read, array write) {
    json params = {
        {"file", file},
        {"read", read},
        {"write", write},
    };
    json headers = {
        {"content-type", "multipart/form-data"},
    };
    string path = "/storage/files";
    size_t pos;
    string searchString;

    return this->client->call(Client::METHOD_POST, path, headers, params);
}

/*
 * Get File
 *
     * Get a file by its unique ID. This endpoint response returns a JSON object
     * with the file metadata.
 * @param string fileId
 * @throws AppwriteException
 * @return array
 */
json Storage::getFile(string fileId) {
    json params = {
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_GET, path, headers, params);
}

/*
 * Update File
 *
     * Update a file by its unique ID. Only users with write permissions have
     * access to update this resource.
 * @param string fileId
 * @param array read
 * @param array write
 * @throws AppwriteException
 * @return array
 */
json Storage::updateFile(string fileId, array read, array write) {
    json params = {
        {"read", read},
        {"write", write},
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_PUT, path, headers, params);
}

/*
 * Delete File
 *
     * Delete a file by its unique ID. Only users with write permissions have
     * access to delete this resource.
 * @param string fileId
 * @throws AppwriteException
 * @return array
 */
json Storage::deleteFile(string fileId) {
    json params = {
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_DELETE, path, headers, params);
}

/*
 * Get File for Download
 *
     * Get a file content by its unique ID. The endpoint response return with a
     * 'Content-Disposition: attachment' header that tells the browser to start
     * downloading the file to user downloads directory.
 * @param string fileId
 * @throws AppwriteException
 * @return string
 */
string Storage::getFileDownload(string fileId) {
    json params = {
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}/download";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_GET, path, headers, params);
}

/*
 * Get File Preview
 *
     * Get a file preview image. Currently, this method supports preview for image
     * files (jpg, png, and gif), other supported formats, like pdf, docs, slides,
     * and spreadsheets, will return the file icon image. You can also pass query
     * string arguments for cutting and resizing your preview image.
 * @param string fileId
 * @param int width
 * @param int height
 * @param string gravity
 * @param int quality
 * @param int borderWidth
 * @param string borderColor
 * @param int borderRadius
 * @param double opacity
 * @param int rotation
 * @param string background
 * @param string output
 * @throws AppwriteException
 * @return string
 */
string Storage::getFilePreview(string fileId, int width, int height, string gravity, int quality, int borderWidth, string borderColor, int borderRadius, double opacity, int rotation, string background, string output) {
    json params = {
        {"width", width},
        {"height", height},
        {"gravity", gravity},
        {"quality", quality},
        {"borderWidth", borderWidth},
        {"borderColor", borderColor},
        {"borderRadius", borderRadius},
        {"opacity", opacity},
        {"rotation", rotation},
        {"background", background},
        {"output", output},
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}/preview";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_GET, path, headers, params);
}

/*
 * Get File for View
 *
     * Get a file content by its unique ID. This endpoint is similar to the
     * download method but returns with no  'Content-Disposition: attachment'
     * header.
 * @param string fileId
 * @throws AppwriteException
 * @return string
 */
string Storage::getFileView(string fileId) {
    json params = {
    };
    json headers = {
        {"content-type", "application/json"},
    };
    string path = "/storage/files/{fileId}/view";
    size_t pos;
    string searchString;
    searchString = "{fileId}";
    pos = path.find(searchString);
    if (pos != std::string::npos)
        path = path.replace(pos, searchString.length(), "fileId");

    return this->client->call(Client::METHOD_GET, path, headers, params);
}
} // namespace Appwrite

File: examples/node/lib/services/storage.js
const Service = require('../service.js');
const AppwriteException = require('../exception.js');

class Storage extends Service {

    /**
     * List Files
     *
     * Get a list of all the user files. You can use the query params to filter
     * your results. On admin mode, this endpoint will return a list of all of the
     * project's files. [Learn more about different API modes](/docs/admin).
     *
     * @param {string} search
     * @param {number} limit
     * @param {number} offset
     * @param {string} orderType
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async listFiles(search, limit, offset, orderType) {
        let path = '/storage/files';
        let payload = {};

        if (typeof search !== 'undefined') {
            payload['search'] = search;
        }

        if (typeof limit !== 'undefined') {
            payload['limit'] = limit;
        }

        if (typeof offset !== 'undefined') {
            payload['offset'] = offset;
        }

        if (typeof orderType !== 'undefined') {
            payload['orderType'] = orderType;
        }

        return await this.client.call('get', path, {
            'content-type': 'application/json',
        }, payload);
    }

    /**
     * Create File
     *
     * Create a new file. The user who creates the file will automatically be
     * assigned to read and write access unless he has passed custom values for
     * read and write arguments.
     *
     * @param {File} file
     * @param {string[]} read
     * @param {string[]} write
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async createFile(file, read, write) {
        if (typeof file === 'undefined') {
            throw new AppwriteException('Missing required parameter: "file"');
        }

        let path = '/storage/files';
        let payload = {};

        if (typeof file !== 'undefined') {
            payload['file'] = file;
        }

        if (typeof read !== 'undefined') {
            payload['read'] = read;
        }

        if (typeof write !== 'undefined') {
            payload['write'] = write;
        }

        return await this.client.call('post', path, {
            'content-type': 'multipart/form-data',
        }, payload);
    }

    /**
     * Get File
     *
     * Get a file by its unique ID. This endpoint response returns a JSON object
     * with the file metadata.
     *
     * @param {string} fileId
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async getFile(fileId) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        let path = '/storage/files/{fileId}'.replace('{fileId}', fileId);
        let payload = {};

        return await this.client.call('get', path, {
            'content-type': 'application/json',
        }, payload);
    }

    /**
     * Update File
     *
     * Update a file by its unique ID. Only users with write permissions have
     * access to update this resource.
     *
     * @param {string} fileId
     * @param {string[]} read
     * @param {string[]} write
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async updateFile(fileId, read, write) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        if (typeof read === 'undefined') {
            throw new AppwriteException('Missing required parameter: "read"');
        }

        if (typeof write === 'undefined') {
            throw new AppwriteException('Missing required parameter: "write"');
        }

        let path = '/storage/files/{fileId}'.replace('{fileId}', fileId);
        let payload = {};

        if (typeof read !== 'undefined') {
            payload['read'] = read;
        }

        if (typeof write !== 'undefined') {
            payload['write'] = write;
        }

        return await this.client.call('put', path, {
            'content-type': 'application/json',
        }, payload);
    }

    /**
     * Delete File
     *
     * Delete a file by its unique ID. Only users with write permissions have
     * access to delete this resource.
     *
     * @param {string} fileId
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async deleteFile(fileId) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        let path = '/storage/files/{fileId}'.replace('{fileId}', fileId);
        let payload = {};

        return await this.client.call('delete', path, {
            'content-type': 'application/json',
        }, payload);
    }

    /**
     * Get File for Download
     *
     * Get a file content by its unique ID. The endpoint response return with a
     * 'Content-Disposition: attachment' header that tells the browser to start
     * downloading the file to user downloads directory.
     *
     * @param {string} fileId
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async getFileDownload(fileId) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        let path = '/storage/files/{fileId}/download'.replace('{fileId}', fileId);
        let payload = {};

        return await this.client.call('get', path, {
            'content-type': 'application/json',
        }, payload, 'arraybuffer');
    }

    /**
     * Get File Preview
     *
     * Get a file preview image. Currently, this method supports preview for image
     * files (jpg, png, and gif), other supported formats, like pdf, docs, slides,
     * and spreadsheets, will return the file icon image. You can also pass query
     * string arguments for cutting and resizing your preview image.
     *
     * @param {string} fileId
     * @param {number} width
     * @param {number} height
     * @param {string} gravity
     * @param {number} quality
     * @param {number} borderWidth
     * @param {string} borderColor
     * @param {number} borderRadius
     * @param {number} opacity
     * @param {number} rotation
     * @param {string} background
     * @param {string} output
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async getFilePreview(fileId, width, height, gravity, quality, borderWidth, borderColor, borderRadius, opacity, rotation, background, output) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        let path = '/storage/files/{fileId}/preview'.replace('{fileId}', fileId);
        let payload = {};

        if (typeof width !== 'undefined') {
            payload['width'] = width;
        }

        if (typeof height !== 'undefined') {
            payload['height'] = height;
        }

        if (typeof gravity !== 'undefined') {
            payload['gravity'] = gravity;
        }

        if (typeof quality !== 'undefined') {
            payload['quality'] = quality;
        }

        if (typeof borderWidth !== 'undefined') {
            payload['borderWidth'] = borderWidth;
        }

        if (typeof borderColor !== 'undefined') {
            payload['borderColor'] = borderColor;
        }

        if (typeof borderRadius !== 'undefined') {
            payload['borderRadius'] = borderRadius;
        }

        if (typeof opacity !== 'undefined') {
            payload['opacity'] = opacity;
        }

        if (typeof rotation !== 'undefined') {
            payload['rotation'] = rotation;
        }

        if (typeof background !== 'undefined') {
            payload['background'] = background;
        }

        if (typeof output !== 'undefined') {
            payload['output'] = output;
        }

        return await this.client.call('get', path, {
            'content-type': 'application/json',
        }, payload, 'arraybuffer');
    }

    /**
     * Get File for View
     *
     * Get a file content by its unique ID. This endpoint is similar to the
     * download method but returns with no  'Content-Disposition: attachment'
     * header.
     *
     * @param {string} fileId
     * @throws {AppwriteException}
     * @returns {Promise}
     */
    async getFileView(fileId) {
        if (typeof fileId === 'undefined') {
            throw new AppwriteException('Missing required parameter: "fileId"');
        }

        let path = '/storage/files/{fileId}/view'.replace('{fileId}', fileId);
        let payload = {};

        return await this.client.call('get', path, {
            'content-type': 'application/json',
        }, payload, 'arraybuffer');
    }
}

module.exports = Storage;